glommio/src/channels/sharding.rs

```diff
@@ -0,0 +1,177 @@
+use std::cell::RefCell;
+use std::fmt::{self, Debug, Formatter};
+use std::rc::Rc;
+
+use futures_lite::{Stream, StreamExt};
+
+use crate::channels::channel_mesh::{FullMesh, Senders};
+use crate::task::JoinHandle;
+use crate::{GlommioError, Local, ResourceType, Result};
+
+/// Trait for handling sharded messages
+///
+/// * `msg` - The message to handle.
+/// * `src_shard` - Id of the shard where the msg is from.
+/// * `cur_shard` - Id of the local shard.
+pub trait Handler<T>: Clone {
+    fn handle(&self, msg: T, src_shard: usize, cur_shard: usize);
+}
+
+/// The public interface for sharding
+pub struct Sharded<T: Send, H> {
+    shard: Rc<Shard<T, H>>,
+    consumers: Vec<JoinHandle<()>>,
+    forward_tasks: Vec<JoinHandle<()>>,
+    closed: bool,
+}
+
+impl<T: Send, H> Debug for Sharded<T, H> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "Sharded")
+    }
+}
+
+/// Alias for sharding function
+pub type ShardFn<T> = fn(&T, usize) -> usize;
+
+impl<T: Send + 'static, H: Handler<T> + 'static> Sharded<T, H> {
+    /// Join a full mesh for sharding
+    pub async fn new(mesh: FullMesh<T>, shard_fn: ShardFn<T>, handler: H) -> Result<Self, ()> {
+        let nr_shards = mesh.nr_peers();
+
+        let (senders, mut receivers) = mesh.join().await?;
+
+        let shard = Rc::new(Shard {
+            nr_shards,
+            shard_id: senders.peer_id(),
+            shard_fn,
+            senders: RefCell::new(Some(senders)),
+            handler: handler.clone(),
+        });
+
+        let mut forward_tasks = Vec::with_capacity(nr_shards);
+        for (src_shard, stream) in receivers.streams() {
+            let handler = handler.clone();
+            let cur_shard = shard.shard_id;
+            let consumer = Local::local(async move {
+                while let Some(msg) = stream.recv().await {
+                    handler.handle(msg, src_shard, cur_shard)
+                }
+            });
+            forward_tasks.push(consumer.detach());
+        }
+
+        Ok(Self {
+            shard,
+            consumers: Vec::new(),
+            forward_tasks,
+            closed: false,
+        })
+    }
+
+    /// Consume messages from a stream. It will return a [`GlommioError`] if this [`Sharded`] is
```

Comment:
those will result in broken links if I am not mistaken.

You need to add `[GlommioError`]: ...` to the end of the help string.
See other instances of that in the shared channels for example.

It is also helpful to run `cargo docs --all-targets` locally, skim through the documentation, and see if the end result is sound.